### PR TITLE
Add HealthIndicator for JavaMailSenderImpl

### LINKED ISSUE #2017: Add email health indicator
To check that the smtp server is reachable.

### PR DESCRIPTION
**Pull request by joshiste**
Closes #2017 

CLA 93620140928055712

### CODE REVIEW

**Comment by snicoll** [comment]
Awesome!

File: spring-boot-actuator/src/main/java/org/springframework/boot/actuate/autoconfigure/HealthIndicatorAutoConfiguration.java [hunk 1]
```diff
@@ -276,4 +280,41 @@ public DiskSpaceHealthIndicatorProperties diskSpaceHealthIndicatorProperties() {
...
+					.entrySet()) {
+				String name = entry.getKey();
+				JavaMailSenderImpl mailSender = entry.getValue();
+				composite.addHealthIndicator(name,
+						createMailSenderHealthIndicator(mailSender));
+			}
+			return composite;
+		}
+
+		private MailHealthIndicator createMailSenderHealthIndicator(
```
Comment (snicoll) [inline spring-boot-actuator/src/main/java/org/springframework/boot/actuate/autoconfigure/HealthIndicatorAutoConfiguration.java]: Looks like a code refactoring that was half applied. should probably be `createMailHealthIndicator`

File: spring-boot-actuator/src/main/java/org/springframework/boot/actuate/health/MailHealthIndicator.java [hunk 2]
```diff
@@ -0,0 +1,50 @@
...
+	}
+
+	@Override
+	protected void doHealthCheck(Builder builder) throws Exception {
+		URI uri = new URI(this.mailSender.getProtocol(), null, this.mailSender.getHost(),
+				this.mailSender.getPort(), null, null, null);
+		builder.withDetail("connection", uri.toString());
+		try {
+			this.mailSender.getSession().getTransport().connect();
+			builder.up();
```
Comment (snicoll) [inline spring-boot-actuator/src/main/java/org/springframework/boot/actuate/health/MailHealthIndicator.java]: Well, that's bad. The `Transport` instance is never closed. `JavaMailSenderImpl#connectTransport` looks like a good fit for what you're doing (relying on the plain `connect` does not seem such a great idea either). Unfortunately it's protected but not rocket science either.

I guess you need a `Transport#close()` class in a guarded finally block code.

File: spring-boot-actuator/src/test/java/org/springframework/boot/actuate/autoconfigure/HealthIndicatorAutoConfigurationTests.java [hunk 3]
```diff
@@ -304,4 +306,37 @@ public DataSource dataSource() {
...
+		this.context.register(MailSenderAutoConfiguration.class,
+				ManagementServerProperties.class, HealthIndicatorAutoConfiguration.class);
+		this.context.refresh();
+
+		Map<String, HealthIndicator> beans = this.context
+				.getBeansOfType(HealthIndicator.class);
+		assertEquals(1, beans.size());
+		assertEquals(ApplicationHealthIndicator.class, beans.values().iterator().next()
+				.getClass());
+	}
```
Comment (snicoll) [inline spring-boot-actuator/src/test/java/org/springframework/boot/actuate/autoconfigure/HealthIndicatorAutoConfigurationTests.java]: You need to add a test where the actual mail breaks down. This should involve a mock on `JavaMailSenderImpl`. Throwing some exception on the `getSession` call would be enough to prove the exception is properly handled. If you want to do things properly, I'd test a failure on `Transport#close`.

File: spring-boot-autoconfigure/src/main/java/org/springframework/boot/autoconfigure/mail/MailSenderAutoConfiguration.java [hunk 4]
```diff
@@ -51,7 +50,7 @@
 	MailProperties properties;
 
 	@Bean
-	public JavaMailSender mailSender() {
+	public JavaMailSenderImpl mailSender() {
```
Comment (snicoll) [inline spring-boot-autoconfigure/src/main/java/org/springframework/boot/autoconfigure/mail/MailSenderAutoConfiguration.java]: I get why you want that but I'd like to commit that separately. I'll push that to master in a minute, please rebase.

File: spring-boot-actuator/src/main/java/org/springframework/boot/actuate/autoconfigure/HealthIndicatorAutoConfiguration.java [hunk 5]
```diff
@@ -276,4 +280,41 @@ public DiskSpaceHealthIndicatorProperties diskSpaceHealthIndicatorProperties() {
 
 	}
 
+	@Configuration
+	@ConditionalOnBean(JavaMailSenderImpl.class)
+	@ConditionalOnProperty(prefix = "management.health.mail", name = "enabled", matchIfMissing = true)
```
Comment (snicoll) [inline spring-boot-actuator/src/main/java/org/springframework/boot/actuate/autoconfigure/HealthIndicatorAutoConfiguration.java]: You need to add a meta-data entry for this key in `META-INF/additional-spring-configuration-metadata.json`. Look at the others for an example.

**Comment by joshiste** [comment]
Thank's for your feedback. I did a forced push. Please have a look.

File: spring-boot-autoconfigure/src/main/java/org/springframework/boot/autoconfigure/mail/MailSenderAutoConfiguration.java (hunk 4 above)
Comment (joshiste) [inline spring-boot-autoconfigure/src/main/java/org/springframework/boot/autoconfigure/mail/MailSenderAutoConfiguration.java]: After reading the discussion in #2536 i didn't see any problems with this.

Comment (snicoll) [inline spring-boot-autoconfigure/src/main/java/org/springframework/boot/autoconfigure/mail/MailSenderAutoConfiguration.java]: There isn't. I prefer a separate commit for that. Don't worry about it.

**Comment by snicoll** [comment]
Alright, this is merged. Thanks a lot.

For a reference, I made some change, please review [this commit](https://github.com/snicoll/spring-boot/commit/935c4ec4ed540f82bfa829ac2513cd1a89aa7663)

**Pull request closed without merging**